File: base64/main.go
// Package base64 holds the functions to base64 encode a string
package base64

var intToLetters map[int]byte

const PADDING = '='

func init() {
	intToLetters = map[int]byte{
		0:  'A',
		1:  'B',
		2:  'C',
		3:  'D',
		4:  'E',
		5:  'F',
		6:  'G',
		7:  'H',
		8:  'I',
		9:  'J',
		10: 'K',
		11: 'L',
		12: 'M',
		13: 'N',
		14: 'O',
		15: 'P',
		16: 'Q',
		17: 'R',
		18: 'S',
		19: 'T',
		20: 'U',
		21: 'V',
		22: 'W',
		23: 'X',
		24: 'Y',
		25: 'Z',
		26: 'a',
		27: 'b',
		28: 'c',
		29: 'd',
		30: 'e',
		31: 'f',
		32: 'g',
		33: 'h',
		34: 'i',
		35: 'j',
		36: 'k',
		37: 'l',
		38: 'm',
		39: 'n',
		40: 'o',
		41: 'p',
		42: 'q',
		43: 'r',
		44: 's',
		45: 't',
		46: 'u',
		47: 'v',
		48: 'w',
		49: 'x',
		50: 'y',
		51: 'z',
		52: '0',
		53: '1',
		54: '2',
		55: '3',
		56: '4',
		57: '5',
		58: '6',
		59: '7',
		60: '8',
		61: '9',
		62: '-', // (minus)
		63: '_', // (underline)
		64: '=', // (pad)
	}
}

// f -> []byte{0,1,1,0,0,1,1,0}
func intToBits(i int32) []byte {
	a := byte(i >> 7 & 0x1)
	b := byte(i >> 6 & 0x1)
	c := byte(i >> 5 & 0x1)
	d := byte(i >> 4 & 0x1)
	e := byte(i >> 3 & 0x1)
	f := byte(i >> 2 & 0x1)
	g := byte(i >> 1 & 0x1)
	h := byte(i >> 0 & 0x1)

	return []byte{a, b, c, d, e, f, g, h}
}

func Base64Encode(str string) string {
	bitWidth := 6

	var bits []byte

	// get all the bits that make up the string
	for _, s := range str {
		bits = append(bits, intToBits(s)...)
	}

	if len(bits) == 0 {
		return ""
	}

	// add padding if needed
	paddingNeeded := bitWidth - (len(bits) % bitWidth)
	if paddingNeeded == bitWidth {
		paddingNeeded = 0
	}

	for range paddingNeeded {
		bits = append(bits, 0)
	}

	var letters []byte

	for i := 0; i <= len(bits)-bitWidth; i += bitWidth {
		idx := (int(bits[i]) << 5)
		idx += (int(bits[i+1]) << 4)
		idx += (int(bits[i+2]) << 3)
		idx += (int(bits[i+3]) << 2)
		idx += (int(bits[i+4]) << 1)
		idx += (int(bits[i+5]) << 0)

		letter := intToLetters[idx]

		letters = append(letters, letter)
	}

	// two padding digits equal a single padding character
	for range paddingNeeded / 2 {
		letters = append(letters, PADDING)
	}

	return string(letters)
}

File: base64/zig/src/main.zig
const std = @import("std");
const based_zig = @import("based_zig");
const Allocator = std.mem.Allocator;

const int_to_letters = [65]u8{
    'A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J',
    'K', 'L', 'M', 'N', 'O', 'P', 'Q', 'R', 'S', 'T',
    'U', 'V', 'W', 'X', 'Y', 'Z',
    'a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j',
    'k', 'l', 'm', 'n', 'o', 'p', 'q', 'r', 's', 't',
    'u', 'v', 'w', 'x', 'y', 'z',
    '0', '1', '2', '3', '4', '5', '6', '7', '8', '9',
    '-', '_', '=',
};

const PADDING: u8 = '=';

// f -> []byte{0,1,1,0,0,1,1,0}
pub fn intToBits(i: u8) [8]u8 {
    const a: u8 = i >> 7 & 0x1;
    const b: u8 = i >> 6 & 0x1;
    const c: u8 = i >> 5 & 0x1;
    const d: u8 = i >> 4 & 0x1;
    const e: u8 = i >> 3 & 0x1;
    const f: u8 = i >> 2 & 0x1;
    const g: u8 = i >> 1 & 0x1;
    const h: u8 = i >> 0 & 0x1;

    return [8]u8{a, b, c, d, e, f, g, h};
}

test "intToBits Test" {
    const res: [8]u8 = intToBits('f');

    try std.testing.expectEqual([8]u8{0, 1, 1, 0, 0, 1, 1, 0}, res);
}

pub fn base64Encode(allocator: std.mem.Allocator, chars: []const u8) !std.ArrayList(u8) {
    const bitWidth: usize = 6;

    var bits: std.ArrayList(u8) = .empty;
    defer bits.deinit(allocator);

    for (chars) |char| {
        const letterBits:[8]u8 = intToBits(char);

        for (letterBits) |letterBit| {
            try bits.append(allocator, letterBit); 
        }
    }

    if (bits.items.len == 0) {
        return bits;
    }

    var paddingNeeded: usize = bitWidth - (bits.items.len % bitWidth);

    if (paddingNeeded == bitWidth) {
        paddingNeeded = 0;
    }

    for (0..paddingNeeded) |_| {
       try bits.append(allocator, 0);
    }

    var letters: std.ArrayList(u8) = .empty;

    var i: usize = 0;

    while (i <= bits.items.len - bitWidth) : (i += bitWidth) {
        var idx: u8 = bits.items[i] << 5;
        idx += bits.items[i+1] << 4;
        idx += bits.items[i+2] << 3;
        idx += bits.items[i+3] << 2;
        idx += bits.items[i+4] << 1;
        idx += bits.items[i+5] << 0;

        const letter: u8 = int_to_letters[idx];

        try letters.append(allocator, letter); 
    }

    for (0..(paddingNeeded/2)) |_| {
        try letters.append(allocator, PADDING);
    }

    return letters;
}

pub fn main() !void {
    const str = "based";

    const allocator = std.heap.page_allocator;

    var encoded = try base64Encode(allocator, str);
    defer encoded.deinit(allocator);
    std.debug.print("{s}\n", .{encoded.items});
}


test "base64 simple test" {
    const str = "hello";
    const allocator = std.heap.page_allocator;

    var encoded = try base64Encode(allocator, str);
    defer encoded.deinit(allocator);

    try std.testing.expectEqualStrings("aGVsbG8=", encoded.items);
}

test "base64 complex test" {
    const str = "hello, this is going to be some sort of a larger string for it to try to do";
    const allocator = std.heap.page_allocator;

    var encoded = try base64Encode(allocator, str);
    defer encoded.deinit(allocator);

    try std.testing.expectEqualStrings("aGVsbG8sIHRoaXMgaXMgZ29pbmcgdG8gYmUgc29tZSBzb3J0IG9mIGEgbGFyZ2VyIHN0cmluZyBmb3IgaXQgdG8gdHJ5IHRvIGRv", encoded.items);
}
